processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -468,39 +479,71 @@ public static void convertV8toV9(File v8Dir, File v9Dir) throws IOException
-            singleValCol = VSizeIndexedInts.fromList(
-                new AbstractList<Integer>()
-                {
-                  @Override
-                  public Integer get(int index)
-                  {
-                    final VSizeIndexedInts ints = finalMultiValCol.get(index);
-                    return ints.size() == 0 ? 0 : ints.get(0) + (bumpedDictionary ? 1 : 0);
-                  }
-
-                  @Override
-                  public int size()
-                  {
-                    return finalMultiValCol.size();
-                  }
-                },
-                dictionary.size()
-            );
+            singleValCol = new AbstractList<Integer>()
+            {
+              @Override
+              public Integer get(int index)
+              {
+                final VSizeIndexedInts ints = finalMultiValCol.get(index);
+                return ints.size() == 0 ? 0 : ints.get(0) + (bumpedDictionary ? 1 : 0);
+              }
+
+              @Override
+              public int size()
+              {
+                return finalMultiValCol.size();
+              }
+            };
+
             multiValCol = null;
           } else {
             builder.setHasMultipleValues(true);
           }
 
-          builder.addSerde(
-              new DictionaryEncodedColumnPartSerde(
-                  dictionary,
-                  singleValCol,
-                  multiValCol,
-                  bitmapSerdeFactory,
-                  bitmaps,
-                  spatialIndex
-              )
-          );
+          final CompressedObjectStrategy.CompressionStrategy compression = indexSpec.getDimensionCompression();
+
+          if (singleValCol != null) {
+            if (compression != null) {
+              builder.addSerde(
+                  DictionaryEncodedColumnPartSerde.createCompressedSingleValue(
+                      dictionary,
+                      CompressedVSizeIntsIndexedSupplier.fromList(
+                          singleValCol,
+                          dictionary.size(),
+                          CompressedVSizeIntsIndexedSupplier.maxIntsInBufferForValue(dictionary.size()),
+                          BYTE_ORDER,
+                          compression
+                      ),
+                      bitmapSerdeFactory,
+                      bitmaps,
+                      spatialIndex,
+                      BYTE_ORDER
+                  )
+              );
+            } else {
+              builder.addSerde(
+                  DictionaryEncodedColumnPartSerde.createUncompressedSingleValue(
+                      dictionary,
+                      VSizeIndexedInts.fromList(singleValCol, dictionary.size()),
+                      bitmapSerdeFactory,
+                      bitmaps,
+                      spatialIndex,
+                      BYTE_ORDER
+                  )
+              );
+            }
+          } else {
+            builder.addSerde(
```

Comment:
My original proposal was only for single value dimensions, I updated the PR description to reflect that.